TPC: move PadFlags and related classes to TPCBaseRecSim

ktf opened the pull request:
o2::tpc::PadFlags and in particular vectors of thereof are affected by an old ROOT
bug in serializing `std::vector<enum struct : short>`.

Because of this we have a custom streamer which needs to be initialised very early in order
to function correctly.

This is fine, however, due to the fact we invoke `TClass::GetClass()` too early, ROOT is
forced a bunch of extra stuff, unneededly, resulting in much increased memory footprint
in analysis, which happens to use DataFormatsTPC.

This makes sure the custom streamer is not initialised statically by DataFormatsTPC and
prevents ROOT from loading the kitchen sink when the TClass::GetClass is invoked too early.

alibuild commented:
Error while checking build/O2/fullCI_slc9 for 35c1e8826ff5bf4671970d56ba0d2e587cdd6b08 at 2025-12-11 23:11:
```
## sw/BUILD/QualityControl-latest/log
/sw/SOURCES/QualityControl/v1.185.0/v1.185.0/Modules/TPC/src/CalDetPublisher.cxx:18:10: fatal error: TPCBase/Painter.h: No such file or directory
/sw/SOURCES/QualityControl/v1.185.0/v1.185.0/Modules/TPC/src/Clusters.cxx:21:10: fatal error: TPCBase/Painter.h: No such file or directory
/sw/SOURCES/QualityControl/v1.185.0/v1.185.0/Modules/TPC/src/LaserTracks.cxx:18:10: fatal error: TPCBase/Painter.h: No such file or directory
/sw/SOURCES/QualityControl/v1.185.0/v1.185.0/Modules/TPC/src/ClusterVisualizer.cxx:18:10: fatal error: TPCBase/Painter.h: No such file or directory
/sw/SOURCES/QualityControl/v1.185.0/v1.185.0/Modules/TPC/src/RawDigits.cxx:20:10: fatal error: TPCBase/Painter.h: No such file or directory
/sw/SOURCES/QualityControl/v1.185.0/v1.185.0/Modules/TPC/src/Utility.cxx:23:10: fatal error: TPCBase/Painter.h: No such file or directory
/sw/SOURCES/QualityControl/v1.185.0/v1.185.0/Modules/TPC/src/IDCs.cxx:18:10: fatal error: TPCBase/CDBInterface.h: No such file or directory
/sw/SOURCES/QualityControl/v1.185.0/v1.185.0/Modules/TPC/src/IDCsVsSACs.cxx:18:10: fatal error: TPCBase/CDBInterface.h: No such file or directory
/sw/SOURCES/QualityControl/v1.185.0/v1.185.0/Modules/TPC/src/SACs.cxx:18:10: fatal error: TPCBase/CDBInterface.h: No such file or directory
ninja: build stopped: cannot make progress due to previous errors.


## sw/BUILD/o2checkcode-latest/log
--
========== List of errors found ==========
++ GRERR=0
++ grep -v clang-diagnostic-error error-log.txt
++ grep ' error:'
grep: error-log.txt: binary file matches
++ GRERR=1
++ [[ 1 == 0 ]]
++ mkdir -p /sw/INSTALLROOT/590369407955fae4cc5f88307f03cce665cad258/slc9_x86-64/o2checkcode/1.0-local150/etc/modulefiles
++ cat
--
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/14913/35c1e8826ff5bf4671970d56ba0d2e587cdd6b08/build_O2_fullCI_slc9/pretty.html).

sawenzel commented:
For my understanding, in which situation will this help and what do you mean by kitchen sink? Isn't the static ROOT variable loaded in TPCBase (versus DataFormatsTPC)?

ktf commented:
Macros fixed to use the new TPCCore package when appropriate.

To be clear, the code is just moved to avoid having it explode memory in analysis. For reco / sim / calibration the change should be simply a mechanical one.

davidrohr commented:
If this is still causing trouble with the custom streamer, we could reconsider switching to a v2 object, and get rid of the streamer?

ktf commented:
v2 is fine for me as long as no one asks to build dozens of old releases to read new data with old release series.

ktf commented:
actually #14918 might be all we need to drop the dependency on DataFormatsTPC in O2Physics.

sawenzel commented:
#14918 looks indeed much easier more convenient

ktf commented:
Indeed. However we might still have a dependency on DataFormatsTPC via one of the other DataFormats... I am trying to do a build.

ktf commented:
#14918 is actually a good cleanup, but the dependency comes in from a different place.. So we do need this.

alibuild commented:
Error while checking build/O2/fullCI_slc9 for 697219108611ec1230e979d0aa3ac0e2f121e294 at 2025-12-17 12:15:
```
## sw/BUILD/QualityControl-latest/log
/sw/SOURCES/QualityControl/v1.185.0/v1.185.0/Modules/TPC/src/CalDetPublisher.cxx:18:10: fatal error: TPCBase/Painter.h: No such file or directory
/sw/SOURCES/QualityControl/v1.185.0/v1.185.0/Modules/TPC/src/Clusters.cxx:21:10: fatal error: TPCBase/Painter.h: No such file or directory
/sw/SOURCES/QualityControl/v1.185.0/v1.185.0/Modules/TPC/src/LaserTracks.cxx:18:10: fatal error: TPCBase/Painter.h: No such file or directory
/sw/SOURCES/QualityControl/v1.185.0/v1.185.0/Modules/TPC/src/ClusterVisualizer.cxx:18:10: fatal error: TPCBase/Painter.h: No such file or directory
/sw/SOURCES/QualityControl/v1.185.0/v1.185.0/Modules/TPC/src/RawDigits.cxx:20:10: fatal error: TPCBase/Painter.h: No such file or directory
/sw/SOURCES/QualityControl/v1.185.0/v1.185.0/Modules/TPC/src/Utility.cxx:23:10: fatal error: TPCBase/Painter.h: No such file or directory
/sw/SOURCES/QualityControl/v1.185.0/v1.185.0/Modules/TPC/src/IDCs.cxx:18:10: fatal error: TPCBase/CDBInterface.h: No such file or directory
/sw/SOURCES/QualityControl/v1.185.0/v1.185.0/Modules/TPC/src/IDCsVsSACs.cxx:18:10: fatal error: TPCBase/CDBInterface.h: No such file or directory
/sw/SOURCES/QualityControl/v1.185.0/v1.185.0/Modules/TPC/src/SACs.cxx:18:10: fatal error: TPCBase/CDBInterface.h: No such file or directory
ninja: build stopped: cannot make progress due to previous errors.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/14913/697219108611ec1230e979d0aa3ac0e2f121e294/build_O2_fullCI_slc9/pretty.html).

alibuild commented:
Error while checking build/O2/fullCI_slc9 for 92ddb2bc67dfadbe09707f25dc98fb99cf2c8479 at 2025-12-12 18:16:
```
## sw/BUILD/o2codechecker-latest/log
100% tests passed, 0 tests failed out of 1


## sw/BUILD/QualityControl-latest/log
/sw/SOURCES/QualityControl/v1.185.0/v1.185.0/Modules/TPC/src/Clusters.cxx:21:10: fatal error: TPCBase/Painter.h: No such file or directory
/sw/SOURCES/QualityControl/v1.185.0/v1.185.0/Modules/TPC/src/CalDetPublisher.cxx:18:10: fatal error: TPCBase/Painter.h: No such file or directory
/sw/SOURCES/QualityControl/v1.185.0/v1.185.0/Modules/TPC/src/RawDigits.cxx:20:10: fatal error: TPCBase/Painter.h: No such file or directory
/sw/SOURCES/QualityControl/v1.185.0/v1.185.0/Modules/TPC/src/LaserTracks.cxx:18:10: fatal error: TPCBase/Painter.h: No such file or directory
/sw/SOURCES/QualityControl/v1.185.0/v1.185.0/Modules/TPC/src/Utility.cxx:23:10: fatal error: TPCBase/Painter.h: No such file or directory
/sw/SOURCES/QualityControl/v1.185.0/v1.185.0/Modules/TPC/src/ClusterVisualizer.cxx:18:10: fatal error: TPCBase/Painter.h: No such file or directory
/sw/SOURCES/QualityControl/v1.185.0/v1.185.0/Modules/TPC/src/IDCs.cxx:18:10: fatal error: TPCBase/CDBInterface.h: No such file or directory
/sw/SOURCES/QualityControl/v1.185.0/v1.185.0/Modules/TPC/src/IDCsVsSACs.cxx:18:10: fatal error: TPCBase/CDBInterface.h: No such file or directory
/sw/SOURCES/QualityControl/v1.185.0/v1.185.0/Modules/TPC/src/SACs.cxx:18:10: fatal error: TPCBase/CDBInterface.h: No such file or directory
ninja: build stopped: cannot make progress due to previous errors.


## sw/BUILD/o2checkcode-latest/log
--
========== List of errors found ==========
++ GRERR=0
++ grep -v clang-diagnostic-error error-log.txt
++ grep ' error:'
grep: error-log.txt: binary file matches
++ GRERR=1
++ [[ 1 == 0 ]]
++ mkdir -p /sw/INSTALLROOT/057c1ba8c2345d9ba14f9d0b266dc0d59b7171ba/slc9_x86-64/o2checkcode/1.0-local46/etc/modulefiles
++ cat
--
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/14913/92ddb2bc67dfadbe09707f25dc98fb99cf2c8479/build_O2_fullCI_slc9/pretty.html).

ktf commented:
This is confirmed to work in reducing memory usage for Analysis. Now trying something less intrusive, however if it does not work I would recommend merging this.

ktf commented:
AliceO2Group/QualityControl#2630 needed for QualityControl to compile with this.

davidrohr commented:
Just for my understanding, what are we trying to fix with this PR? Compatibility with new O2 and old ROOT?
We went for this custom streamer since we said it is simpler than uploading many new objects, but meanwhile I am not sure anymore.
This is getting more and more complicated, and at some point we should stop creating a mess, and instead just say we do not suppport the old ROOT any more, or is there a reason to support it?
And if we drop the support, at some point we can also drop the streamer and just upload new CCDB objects in the correct format. I propose we discuss this again with all involved people, perhaps at the S&C meeting.

ktf commented:
No, we are trying to work around the fact that with the custom streamer as it is written, ROOT needs to load a bunch of dictionaries which were previously not loaded by simply linking the DataFormatsTPC library. This results in an exploded memory footprint for any analysis / workflow that depends, directly or indirectly, on DataFormatsTPC.

This patch solves the problem by moving the static in a new library (libTPCCore) which is not linked by any package needed by analysis.

As a side remark on this attempt, I have the impression that we could improve O2 memory footprint by a more careful usage of dependencies, in particular, avoiding having a dependency on TCanvas in TPCBase (and elsewhere).

Let's discuss in the meeting.

davidrohr commented:
Ok, but that means it is a fallout of the custom streamer.
In general I absolutely fully agree to reduce dependencies, but I would do this independent from the streamer stuff.

So my opi ion is then again, someone needa to write a script to reupload all the objects, and then let's just get rid of all this custon streames mess.

ktf commented:
I have no objections if the detector experts want to clean their detector specific objects in CCDB and DPG deprecates old releases accessing them.

That said even without the custom streamer, this refactoring makes sense to me. I can simply change the commit message. None of the classes I moved are needed in analysis and they just bring a lot of baggage with them (including TCanvas...).

ktf commented:
Even better would be to get rid of any dependency on detector code in analysis, of course, however that's not something I can personally do in the short term.

ktf commented:
As discussed in the meeting yesterday, waiting for @wiechula to greenlight this (independently of the custom streamer).

davidrohr reviewed (commented):
So if I understand correctly, all this does is to move some classes around from TPCBase to TPCCode, such that one does not need to include TPCBase in so many places?
While I agree in general, I think the names are a bit confusing. Reading just TPCBase and TPCCore, it is not clear to me what is the lower level library. I would assume future additions will end up in a ransom choice of the two.
Is it such that TPCBase should only be used for Reconstruction and Simulation? If yes, than perhaps I would renaname it to TPCBase and TPCBaseRecSim, to make it clear what is the difference.

ktf commented:
I am happy to rename when I am back. Merry Christmas!🎄

wiechula commented:
Is this now superseded by https://github.com/AliceO2Group/AliceO2/pull/14918, or is this restructuring still needed?

ktf commented:
#14918 is an unrelated cleanup. I think it would still be good to have this, as it removes dependency on `Painter.h` / `CDBInterface` from `TPCCore`.

ktf committed:
TPC: move PadFlags and related classes to TPCBaseRecSim

o2::tpc::PadFlags and in particular vectors of thereof are affected by an old ROOT
bug in serializing `std::vector<enum struct : short>`.

Because of this we have a custom streamer which needs to be initialised very early in order
to function correctly.

This is fine, however, due to the fact we invoke `TClass::GetClass()` too early, ROOT is
forced a bunch of extra stuff, unneededly, resulting in much increased memory footprint
in analysis, which happens to use DataFormatsTPC.

This makes sure the custom streamer is not initialised statically by DataFormatsTPC and
prevents ROOT from loading the kitchen sink when the TClass::GetClass is invoked too early.

ktf commented:
@davidrohr updated with the new name `TPCBaseRecSim` in place of the ambiguous `TPCCore`.

alibuild commented:
Error while checking build/O2/fullCI_slc9 for 515e4cb2c3959cd99ef6cdd8f73f81a20ba0c5df at 2026-01-06 11:07:
```
## sw/BUILD/O2-latest/log
CMake Error at cmake/O2TargetRootDictionary.cmake:83 (message):
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/14913/515e4cb2c3959cd99ef6cdd8f73f81a20ba0c5df/build_O2_fullCI_slc9/pretty.html).

alibuild commented:
Error while checking build/O2/fullCI_slc9 for c8d7819e55742e99c7ca41a2877ed6cf1eed3101 at 2026-01-09 14:59:
```
Suppressed 4204 warnings (4204 in non-user code).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
4315 warnings generated.
Suppressed 4315 warnings (4315 in non-user code).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
4101 warnings generated.
Suppressed 4101 warnings (4099 in non-user code, 2 with check filters).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
4590 warnings generated.
Suppressed 4591 warnings (4590 in non-user code, 1 NOLINT).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
4972 warnings generated.
Suppressed 4972 warnings (4972 in non-user code).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
4972 warnings generated.
Suppressed 4972 warnings (4972 in non-user code).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
5035 warnings generated.
Suppressed 5035 warnings (5033 in non-user code, 2 with check filters).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
/sw/slc9_x86-64/Clang/v20.1.7-7/bin-safe/clang-tidy -header-filter=.*SOURCES(?!.*/3rdparty/).* -checks=-*,modernize-avoid-bind,modernize-deprecated-headers,modernize-make-shared,modernize-raw-string-literal,modernize-redundant-void-arg,modernize-replace-auto-ptr,modernize-replace-random-shuffle,modernize-shrink-to-fit,modernize-unary-static-assert,modernize-use-equals-default,modernize-use-noexcept,modernize-use-nullptr,modernize-use-override,modernize-use-transparent-functors,modernize-use-uncaught-exceptions,readability-braces-around-statements --load /sw/slc9_x86-64/o2codechecker/v20.1.7-alice2-local3/lib/libclangTidyAliceO2Module.so --extra-arg=--gcc-install-dir= -p=/sw/BUILD/2bae3e1024869f12eb7d6e8f75e7e27efc898bb3/o2checkcode /sw/SOURCES/O2/14913-slc9_x86-64/0/Detectors/GlobalTracking/src/MatchCosmicsParams.cxx --extra-arg=-fopenmp
/sw/slc9_x86-64/Clang/v20.1.7-7/bin-safe/clang-tidy -header-filter=.*SOURCES(?!.*/3rdparty/).* -checks=-*,modernize-avoid-bind,modernize-deprecated-headers,modernize-make-shared,modernize-raw-string-literal,modernize-redundant-void-arg,modernize-replace-auto-ptr,modernize-replace-random-shuffle,modernize-shrink-to-fit,modernize-unary-static-assert,modernize-use-equals-default,modernize-use-noexcept,modernize-use-nullptr,modernize-use-override,modernize-use-transparent-functors,modernize-use-uncaught-exceptions,readability-braces-around-statements --load /sw/slc9_x86-64/o2codechecker/v20.1.7-alice2-local3/lib/libclangTidyAliceO2Module.so --extra-arg=--gcc-install-dir= -p=/sw/BUILD/2bae3e1024869f12eb7d6e8f75e7e27efc898bb3/o2checkcode /sw/SOURCES/O2/14913-slc9_x86-64/0/Detectors/GlobalTracking/src/MatchGlobalFwd.cxx --extra-arg=-fopenmp
/sw/slc9_x86-64/Clang/v20.1.7-7/bin-safe/clang-tidy -header-filter=.*SOURCES(?!.*/3rdparty/).* -checks=-*,modernize-avoid-bind,modernize-deprecated-headers,modernize-make-shared,modernize-raw-string-literal,modernize-redundant-void-arg,modernize-replace-auto-ptr,modernize-replace-random-shuffle,modernize-shrink-to-fit,modernize-unary-static-assert,modernize-use-equals-default,modernize-use-noexcept,modernize-use-nullptr,modernize-use-override,modernize-use-transparent-functors,modernize-use-uncaught-exceptions,readability-braces-around-statements --load /sw/slc9_x86-64/o2codechecker/v20.1.7-alice2-local3/lib/libclangTidyAliceO2Module.so --extra-arg=--gcc-install-dir= -p=/sw/BUILD/2bae3e1024869f12eb7d6e8f75e7e27efc898bb3/o2checkcode /sw/SOURCES/O2/14913-slc9_x86-64/0/Detectors/GlobalTracking/src/MatchGlobalFwdAssessment.cxx --extra-arg=-fopenmp
/sw/slc9_x86-64/Clang/v20.1.7-7/bin-safe/clang-tidy -header-filter=.*SOURCES(?!.*/3rdparty/).* -checks=-*,modernize-avoid-bind,modernize-deprecated-headers,modernize-make-shared,modernize-raw-string-literal,modernize-redundant-void-arg,modernize-replace-auto-ptr,modernize-replace-random-shuffle,modernize-shrink-to-fit,modernize-unary-static-assert,modernize-use-equals-default,modernize-use-noexcept,modernize-use-nullptr,modernize-use-override,modernize-use-transparent-functors,modernize-use-uncaught-exceptions,readability-braces-around-statements --load /sw/slc9_x86-64/o2codechecker/v20.1.7-alice2-local3/lib/libclangTidyAliceO2Module.so --extra-arg=--gcc-install-dir= -p=/sw/BUILD/2bae3e1024869f12eb7d6e8f75e7e27efc898bb3/o2checkcode /sw/SOURCES/O2/14913-slc9_x86-64/0/Detectors/GlobalTracking/src/MatchGlobalFwdParam.cxx --extra-arg=-fopenmp
/sw/slc9_x86-64/Clang/v20.1.7-7/bin-safe/clang-tidy -header-filter=.*SOURCES(?!.*/3rdparty/).* -checks=-*,modernize-avoid-bind,modernize-deprecated-headers,modernize-make-shared,modernize-raw-string-literal,modernize-redundant-void-arg,modernize-replace-auto-ptr,modernize-replace-random-shuffle,modernize-shrink-to-fit,modernize-unary-static-assert,modernize-use-equals-default,modernize-use-noexcept,modernize-use-nullptr,modernize-use-override,modernize-use-transparent-functors,modernize-use-uncaught-exceptions,readability-braces-around-statements --load /sw/slc9_x86-64/o2codechecker/v20.1.7-alice2-local3/lib/libclangTidyAliceO2Module.so --extra-arg=--gcc-install-dir= -p=/sw/BUILD/2bae3e1024869f12eb7d6e8f75e7e27efc898bb3/o2checkcode /sw/SOURCES/O2/14913-slc9_x86-64/0/Detectors/GlobalTracking/src/MatchTOFParams.cxx --extra-arg=-fopenmp
/sw/slc9_x86-64/Clang/v20.1.7-7/bin-safe/clang-tidy -header-filter=.*SOURCES(?!.*/3rdparty/).* -checks=-*,modernize-avoid-bind,modernize-deprecated-headers,modernize-make-shared,modernize-raw-string-literal,modernize-redundant-void-arg,modernize-replace-auto-ptr,modernize-replace-random-shuffle,modernize-shrink-to-fit,modernize-unary-static-assert,modernize-use-equals-default,modernize-use-noexcept,modernize-use-nullptr,modernize-use-override,modernize-use-transparent-functors,modernize-use-uncaught-exceptions,readability-braces-around-statements --load /sw/slc9_x86-64/o2codechecker/v20.1.7-alice2-local3/lib/libclangTidyAliceO2Module.so --extra-arg=--gcc-install-dir= -p=/sw/BUILD/2bae3e1024869f12eb7d6e8f75e7e27efc898bb3/o2checkcode /sw/SOURCES/O2/14913-slc9_x86-64/0/Detectors/GlobalTracking/src/MatchITSTPCQC.cxx --extra-arg=-fopenmp
/sw/slc9_x86-64/Clang/v20.1.7-7/bin-safe/clang-tidy -header-filter=.*SOURCES(?!.*/3rdparty/).* -checks=-*,modernize-avoid-bind,modernize-deprecated-headers,modernize-make-shared,modernize-raw-string-literal,modernize-redundant-void-arg,modernize-replace-auto-ptr,modernize-replace-random-shuffle,modernize-shrink-to-fit,modernize-unary-static-assert,modernize-use-equals-default,modernize-use-noexcept,modernize-use-nullptr,modernize-use-override,modernize-use-transparent-functors,modernize-use-uncaught-exceptions,readability-braces-around-statements --load /sw/slc9_x86-64/o2codechecker/v20.1.7-alice2-local3/lib/libclangTidyAliceO2Module.so --extra-arg=--gcc-install-dir= -p=/sw/BUILD/2bae3e1024869f12eb7d6e8f75e7e27efc898bb3/o2checkcode /sw/SOURCES/O2/14913-slc9_x86-64/0/Detectors/GlobalTracking/src/ITSTPCMatchingQCParams.cxx --extra-arg=-fopenmp
/sw/slc9_x86-64/Clang/v20.1.7-7/bin-safe/clang-tidy -header-filter=.*SOURCES(?!.*/3rdparty/).* -checks=-*,modernize-avoid-bind,modernize-deprecated-headers,modernize-make-shared,modernize-raw-string-literal,modernize-redundant-void-arg,modernize-replace-auto-ptr,modernize-replace-random-shuffle,modernize-shrink-to-fit,modernize-unary-static-assert,modernize-use-equals-default,modernize-use-noexcept,modernize-use-nullptr,modernize-use-override,modernize-use-transparent-functors,modernize-use-uncaught-exceptions,readability-braces-around-statements --load /sw/slc9_x86-64/o2codechecker/v20.1.7-alice2-local3/lib/libclangTidyAliceO2Module.so --extra-arg=--gcc-install-dir= -p=/sw/BUILD/2bae3e1024869f12eb7d6e8f75e7e27efc898bb3/o2checkcode /sw/SOURCES/O2/14913-slc9_x86-64/0/Detectors/GlobalTrackingWorkflow/src/TrackWriterTPCITSSpec.cxx --extra-arg=-fopenmp
/sw/slc9_x86-64/Clang/v20.1.7-7/bin-safe/clang-tidy -header-filter=.*SOURCES(?!.*/3rdparty/).* -checks=-*,modernize-avoid-bind,modernize-deprecated-headers,modernize-make-shared,modernize-raw-string-literal,modernize-redundant-void-arg,modernize-replace-auto-ptr,modernize-replace-random-shuffle,modernize-shrink-to-fit,modernize-unary-static-assert,modernize-use-equals-default,modernize-use-noexcept,modernize-use-nullptr,modernize-use-override,modernize-use-transparent-functors,modernize-use-uncaught-exceptions,readability-braces-around-statements --load /sw/slc9_x86-64/o2codechecker/v20.1.7-alice2-local3/lib/libclangTidyAliceO2Module.so --extra-arg=--gcc-install-dir= -p=/sw/BUILD/2bae3e1024869f12eb7d6e8f75e7e27efc898bb3/o2checkcode /sw/SOURCES/O2/14913-slc9_x86-64/0/Detectors/GlobalTrackingWorkflow/src/TPCITSMatchingSpec.cxx --extra-arg=-fopenmp
4259 warnings generated.
Suppressed 4259 warnings (4256 in non-user code, 3 with check filters).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
4238 warnings generated.
Suppressed 4238 warnings (4237 in non-user code, 1 with check filters).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
2676 warnings generated.
Suppressed 2677 warnings (2676 in non-user code, 1 NOLINT).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
4969 warnings generated.
Suppressed 4971 warnings (4969 in non-user code, 2 NOLINT).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
4969 warnings generated.
Suppressed 4971 warnings (4969 in non-user code, 2 NOLINT).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
4949 warnings generated.
Suppressed 4951 warnings (4949 in non-user code, 2 NOLINT).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
4957 warnings generated.
Suppressed 4959 warnings (4957 in non-user code, 2 NOLINT).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/14913/c8d7819e55742e99c7ca41a2877ed6cf1eed3101/build_O2_fullCI_slc9/pretty.html).

ktf commented:
The errors are due to the fact QC needs https://github.com/AliceO2Group/QualityControl/pull/2630 for this to work. If there is consensus this cleanup might reduce downstream dependencies, I would go ahead and merge the QC PR so that we can properly test this. If people are not interested, please let me know and I will close both.

wiechula reviewed (approved):
@ktf , the changes are fine for me.

davidrohr commented:
Fine for me, and I think if we want to force merge stuff breaking something temporarily, now is the time :).

ktf commented:
We do not need to force merge, actually. Merging QC first should be backward  - forward compatible.

davidrohr commented:
@ktf : Also with the new QC tag 1.185.1, the QC build fails not finding headers:

```
FAILED: Modules/TPC/CMakeFiles/O2QcTPC.dir/src/Utility.cxx.o 
/sw/slc9_x86-64/GCC-Toolchain/v14.2.0-alice2-local1/bin/c++ -DBOOST_ASIO_HAS_HAS_STD_CHRONO -DBOOST_ATOMIC_DYN_LINK -DBOOST_ATOMIC_NO_LIB -DBOOST_CONTAINER_DYN_LINK -DBOOST_CONTAINER_NO_LIB -DBOOST_ERROR_CODE_HEADER_ONLY -DBOOST_FILESYSTEM_DYN_LINK -DBOOST_FILESYSTEM_NO_LIB -DBOOST_IOSTREAMS_DYN_LINK -DBOOST_IOSTREAMS_NO_LIB -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_PROGRAM_OPTIONS_NO_LIB -DBOOST_REGEX_DYN_LINK -DBOOST_REGEX_NO_LIB -DBOOST_SERIALIZATION_DYN_LINK -DBOOST_SERIALIZATION_NO_LIB -DDPL_ENABLE_BACKTRACE -DFAIRMQ_HAS_STD_FILESYSTEM=1 -DFAIRMQ_HAS_STD_PMR=1 -DFMT_SHARED -DO2QcTPC_EXPORTS -DRANS_ENABLE_JSON -DRANS_OPENMP -DROOT_SUPPORT_CLAD -I/sw/SOURCES/QualityControl/v1.185.1/v1.185.1/Modules/TPC/include -I/sw/SOURCES/QualityControl/v1.185.1/v1.185.1/Modules/TPC/src -I/sw/SOURCES/QualityControl/v1.185.1/v1.185.1/Modules/TPC -I/sw/SOURCES/QualityControl/v1.185.1/v1.185.1/Framework/include -I/sw/slc9_x86-64/moderncppkafka/v2023.03.07-local1/include -I/sw/slc9_x86-64/librdkafka/v2.3.0-local1/include -I/sw/BUILD/9b9a947943d0776e6cc9d4abf198209fd32e9278/QualityControl/Framework -I/sw/SOURCES/QualityControl/v1.185.1/v1.185.1/Modules/Common/include -isystem /sw/slc9_x86-64/O2/14913-slc9_x86-64-local3/include/GPU -isystem /sw/slc9_x86-64/boost/v1.83.0-alice2-local2/include -isystem /sw/slc9_x86-64/FairLogger/v2.3.1-local1/include -isystem /sw/slc9_x86-64/fmt/11.1.2-local1/include -isystem /sw/slc9_x86-64/FairMQ/v1.10.0-local2/include/fairmq -isystem /sw/slc9_x86-64/FairMQ/v1.10.0-local2/include -isystem /sw/slc9_x86-64/ROOT/v6-36-04-alice9-local2/include -isystem /sw/slc9_x86-64/Common-O2/v1.6.3-local2/include -isystem /sw/slc9_x86-64/Monitoring/v3.19.8-local2/include -isystem /sw/slc9_x86-64/Configuration/v2.8.0-local3/include -isystem /sw/slc9_x86-64/Control-OCCPlugin/v1.46.1-local3/include/Occ -isystem /sw/slc9_x86-64/protobuf/v29.3-local1/include -isystem /sw/slc9_x86-64/abseil/20240722.0-local1/include -isystem /sw/slc9_x86-64/grpc/v1.71.0-local1/include -isystem /sw/slc9_x86-64/re2/2024-07-02-local1/include -isystem /sw/slc9_x86-64/zlib/v1.3.1-local1/include -isystem /sw/slc9_x86-64/c-ares/1.18.1-local1/include -isystem /sw/slc9_x86-64/O2/14913-slc9_x86-64-local3/include -isystem /sw/slc9_x86-64/curl/7.70.0-local1/include -isystem /sw/slc9_x86-64/RapidJSON/v1.1.0-alice2-local1/include -isystem /sw/slc9_x86-64/arrow/v20.0.0-alice1-local3/include -isystem /sw/slc9_x86-64/ms_gsl/4.0.0-local1/include -isystem /sw/slc9_x86-64/Clang/v20.1.7-local2/include -isystem /sw/slc9_x86-64/libuv/v1.40.0-local1/include -isystem /sw/slc9_x86-64/Vc/1.4.5-local1/include -isystem /sw/slc9_x86-64/TBB/v2021.5.0-local1/include -isystem /sw/slc9_x86-64/libjalienO2/0.2.0-local1/include -isystem /sw/slc9_x86-64/bookkeeping-api/v1.9.2-local1/include -isystem /sw/slc9_x86-64/libInfoLogger/v2.8.3-local2/include -isystem /sw/slc9_x86-64/FairRoot/v18.4.9-alice3-local4/include -isystem /sw/slc9_x86-64/VMC/v2-1-local3/include/vmc -fPIC -O2 -std=c++20 -Werror -Wno-error=deprecated-declarations -Wno-error=unused-function -O2 -g -DNDEBUG -std=gnu++20 -fPIC -fopenmp -MD -MT Modules/TPC/CMakeFiles/O2QcTPC.dir/src/Utility.cxx.o -MF Modules/TPC/CMakeFiles/O2QcTPC.dir/src/Utility.cxx.o.d -o Modules/TPC/CMakeFiles/O2QcTPC.dir/src/Utility.cxx.o -c /sw/SOURCES/QualityControl/v1.185.1/v1.185.1/Modules/TPC/src/Utility.cxx
/sw/SOURCES/QualityControl/v1.185.1/v1.185.1/Modules/TPC/src/Utility.cxx:24:10: fatal error: TPCBase/Painter.h: No such file or directory
   24 | #include "TPCBase/Painter.h"
      |          ^~~~~~~~~~~~~~~~~~~
compilation terminated.
[367/631] Building CXX object Modules/MUON/Common/CMakeFiles/O2QcMUONCommon.dir/src/TracksPostProcessingConfig.cxx.o
[368/631] Building CXX object Modules/HMPID/CMakeFiles/O2QcHMPID.dir/src/HmpidTaskMatches.cxx.o
[369/631] Building CXX object Modules/MUON/Common/CMakeFiles/O2QcMUONCommon.dir/src/MuonTrack.cxx.o
[370/631] Building CXX object Modules/Daq/CMakeFiles/O2QcDaq.dir/src/DaqTask.cxx.o
[371/631] Generating G__O2QcFDD.cxx, ../../../lib/G__O2QcFDD_rdict.pcm, ../../../lib/libO2QcFDD.rootmap
[372/631] Building CXX object Modules/ZDC/CMakeFiles/O2QcZDC.dir/src/ZDCRawDataCheck.cxx.o
[373/631] Building CXX object Modules/PID/CMakeFiles/O2QcPID.dir/G__O2QcPID.cxx.o
[374/631] Building CXX object Modules/TPC/CMakeFiles/O2QcTPC.dir/src/ROCReductor.cxx.o
[375/631] Generating G__O2QcFT0.cxx, ../../../lib/G__O2QcFT0_rdict.pcm, ../../../lib/libO2QcFT0.rootmap
[376/631] Linking CXX shared library lib/libO2QcDaq.so
[377/631] Building CXX object Modules/TPC/CMakeFiles/O2QcTPC.dir/src/IDCsVsSACs.cxx.o
FAILED: Modules/TPC/CMakeFiles/O2QcTPC.dir/src/IDCsVsSACs.cxx.o 
/sw/slc9_x86-64/GCC-Toolchain/v14.2.0-alice2-local1/bin/c++ -DBOOST_ASIO_HAS_HAS_STD_CHRONO -DBOOST_ATOMIC_DYN_LINK -DBOOST_ATOMIC_NO_LIB -DBOOST_CONTAINER_DYN_LINK -DBOOST_CONTAINER_NO_LIB -DBOOST_ERROR_CODE_HEADER_ONLY -DBOOST_FILESYSTEM_DYN_LINK -DBOOST_FILESYSTEM_NO_LIB -DBOOST_IOSTREAMS_DYN_LINK -DBOOST_IOSTREAMS_NO_LIB -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_PROGRAM_OPTIONS_NO_LIB -DBOOST_REGEX_DYN_LINK -DBOOST_REGEX_NO_LIB -DBOOST_SERIALIZATION_DYN_LINK -DBOOST_SERIALIZATION_NO_LIB -DDPL_ENABLE_BACKTRACE -DFAIRMQ_HAS_STD_FILESYSTEM=1 -DFAIRMQ_HAS_STD_PMR=1 -DFMT_SHARED -DO2QcTPC_EXPORTS -DRANS_ENABLE_JSON -DRANS_OPENMP -DROOT_SUPPORT_CLAD -I/sw/SOURCES/QualityControl/v1.185.1/v1.185.1/Modules/TPC/include -I/sw/SOURCES/QualityControl/v1.185.1/v1.185.1/Modules/TPC/src -I/sw/SOURCES/QualityControl/v1.185.1/v1.185.1/Modules/TPC -I/sw/SOURCES/QualityControl/v1.185.1/v1.185.1/Framework/include -I/sw/slc9_x86-64/moderncppkafka/v2023.03.07-local1/include -I/sw/slc9_x86-64/librdkafka/v2.3.0-local1/include -I/sw/BUILD/9b9a947943d0776e6cc9d4abf198209fd32e9278/QualityControl/Framework -I/sw/SOURCES/QualityControl/v1.185.1/v1.185.1/Modules/Common/include -isystem /sw/slc9_x86-64/O2/14913-slc9_x86-64-local3/include/GPU -isystem /sw/slc9_x86-64/boost/v1.83.0-alice2-local2/include -isystem /sw/slc9_x86-64/FairLogger/v2.3.1-local1/include -isystem /sw/slc9_x86-64/fmt/11.1.2-local1/include -isystem /sw/slc9_x86-64/FairMQ/v1.10.0-local2/include/fairmq -isystem /sw/slc9_x86-64/FairMQ/v1.10.0-local2/include -isystem /sw/slc9_x86-64/ROOT/v6-36-04-alice9-local2/include -isystem /sw/slc9_x86-64/Common-O2/v1.6.3-local2/include -isystem /sw/slc9_x86-64/Monitoring/v3.19.8-local2/include -isystem /sw/slc9_x86-64/Configuration/v2.8.0-local3/include -isystem /sw/slc9_x86-64/Control-OCCPlugin/v1.46.1-local3/include/Occ -isystem /sw/slc9_x86-64/protobuf/v29.3-local1/include -isystem /sw/slc9_x86-64/abseil/20240722.0-local1/include -isystem /sw/slc9_x86-64/grpc/v1.71.0-local1/include -isystem /sw/slc9_x86-64/re2/2024-07-02-local1/include -isystem /sw/slc9_x86-64/zlib/v1.3.1-local1/include -isystem /sw/slc9_x86-64/c-ares/1.18.1-local1/include -isystem /sw/slc9_x86-64/O2/14913-slc9_x86-64-local3/include -isystem /sw/slc9_x86-64/curl/7.70.0-local1/include -isystem /sw/slc9_x86-64/RapidJSON/v1.1.0-alice2-local1/include -isystem /sw/slc9_x86-64/arrow/v20.0.0-alice1-local3/include -isystem /sw/slc9_x86-64/ms_gsl/4.0.0-local1/include -isystem /sw/slc9_x86-64/Clang/v20.1.7-local2/include -isystem /sw/slc9_x86-64/libuv/v1.40.0-local1/include -isystem /sw/slc9_x86-64/Vc/1.4.5-local1/include -isystem /sw/slc9_x86-64/TBB/v2021.5.0-local1/include -isystem /sw/slc9_x86-64/libjalienO2/0.2.0-local1/include -isystem /sw/slc9_x86-64/bookkeeping-api/v1.9.2-local1/include -isystem /sw/slc9_x86-64/libInfoLogger/v2.8.3-local2/include -isystem /sw/slc9_x86-64/FairRoot/v18.4.9-alice3-local4/include -isystem /sw/slc9_x86-64/VMC/v2-1-local3/include/vmc -fPIC -O2 -std=c++20 -Werror -Wno-error=deprecated-declarations -Wno-error=unused-function -O2 -g -DNDEBUG -std=gnu++20 -fPIC -fopenmp -MD -MT Modules/TPC/CMakeFiles/O2QcTPC.dir/src/IDCsVsSACs.cxx.o -MF Modules/TPC/CMakeFiles/O2QcTPC.dir/src/IDCsVsSACs.cxx.o.d -o Modules/TPC/CMakeFiles/O2QcTPC.dir/src/IDCsVsSACs.cxx.o -c /sw/SOURCES/QualityControl/v1.185.1/v1.185.1/Modules/TPC/src/IDCsVsSACs.cxx
/sw/SOURCES/QualityControl/v1.185.1/v1.185.1/Modules/TPC/src/IDCsVsSACs.cxx:26:10: fatal error: TPCBaseRecSim/CalDet.h: No such file or directory
   26 | #include "TPCBaseRecSim/CalDet.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
[378/631] Building CXX object Modules/TPC/CMakeFiles/O2QcTPC.dir/src/IDCs.cxx.o
FAILED: Modules/TPC/CMakeFiles/O2QcTPC.dir/src/IDCs.cxx.o 
/sw/slc9_x86-64/GCC-Toolchain/v14.2.0-alice2-local1/bin/c++ -DBOOST_ASIO_HAS_HAS_STD_CHRONO -DBOOST_ATOMIC_DYN_LINK -DBOOST_ATOMIC_NO_LIB -DBOOST_CONTAINER_DYN_LINK -DBOOST_CONTAINER_NO_LIB -DBOOST_ERROR_CODE_HEADER_ONLY -DBOOST_FILESYSTEM_DYN_LINK -DBOOST_FILESYSTEM_NO_LIB -DBOOST_IOSTREAMS_DYN_LINK -DBOOST_IOSTREAMS_NO_LIB -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_PROGRAM_OPTIONS_NO_LIB -DBOOST_REGEX_DYN_LINK -DBOOST_REGEX_NO_LIB -DBOOST_SERIALIZATION_DYN_LINK -DBOOST_SERIALIZATION_NO_LIB -DDPL_ENABLE_BACKTRACE -DFAIRMQ_HAS_STD_FILESYSTEM=1 -DFAIRMQ_HAS_STD_PMR=1 -DFMT_SHARED -DO2QcTPC_EXPORTS -DRANS_ENABLE_JSON -DRANS_OPENMP -DROOT_SUPPORT_CLAD -I/sw/SOURCES/QualityControl/v1.185.1/v1.185.1/Modules/TPC/include -I/sw/SOURCES/QualityControl/v1.185.1/v1.185.1/Modules/TPC/src -I/sw/SOURCES/QualityControl/v1.185.1/v1.185.1/Modules/TPC -I/sw/SOURCES/QualityControl/v1.185.1/v1.185.1/Framework/include -I/sw/slc9_x86-64/moderncppkafka/v2023.03.07-local1/include -I/sw/slc9_x86-64/librdkafka/v2.3.0-local1/include -I/sw/BUILD/9b9a947943d0776e6cc9d4abf198209fd32e9278/QualityControl/Framework -I/sw/SOURCES/QualityControl/v1.185.1/v1.185.1/Modules/Common/include -isystem /sw/slc9_x86-64/O2/14913-slc9_x86-64-local3/include/GPU -isystem /sw/slc9_x86-64/boost/v1.83.0-alice2-local2/include -isystem /sw/slc9_x86-64/FairLogger/v2.3.1-local1/include -isystem /sw/slc9_x86-64/fmt/11.1.2-local1/include -isystem /sw/slc9_x86-64/FairMQ/v1.10.0-local2/include/fairmq -isystem /sw/slc9_x86-64/FairMQ/v1.10.0-local2/include -isystem /sw/slc9_x86-64/ROOT/v6-36-04-alice9-local2/include -isystem /sw/slc9_x86-64/Common-O2/v1.6.3-local2/include -isystem /sw/slc9_x86-64/Monitoring/v3.19.8-local2/include -isystem /sw/slc9_x86-64/Configuration/v2.8.0-local3/include -isystem /sw/slc9_x86-64/Control-OCCPlugin/v1.46.1-local3/include/Occ -isystem /sw/slc9_x86-64/protobuf/v29.3-local1/include -isystem /sw/slc9_x86-64/abseil/20240722.0-local1/include -isystem /sw/slc9_x86-64/grpc/v1.71.0-local1/include -isystem /sw/slc9_x86-64/re2/2024-07-02-local1/include -isystem /sw/slc9_x86-64/zlib/v1.3.1-local1/include -isystem /sw/slc9_x86-64/c-ares/1.18.1-local1/include -isystem /sw/slc9_x86-64/O2/14913-slc9_x86-64-local3/include -isystem /sw/slc9_x86-64/curl/7.70.0-local1/include -isystem /sw/slc9_x86-64/RapidJSON/v1.1.0-alice2-local1/include -isystem /sw/slc9_x86-64/arrow/v20.0.0-alice1-local3/include -isystem /sw/slc9_x86-64/ms_gsl/4.0.0-local1/include -isystem /sw/slc9_x86-64/Clang/v20.1.7-local2/include -isystem /sw/slc9_x86-64/libuv/v1.40.0-local1/include -isystem /sw/slc9_x86-64/Vc/1.4.5-local1/include -isystem /sw/slc9_x86-64/TBB/v2021.5.0-local1/include -isystem /sw/slc9_x86-64/libjalienO2/0.2.0-local1/include -isystem /sw/slc9_x86-64/bookkeeping-api/v1.9.2-local1/include -isystem /sw/slc9_x86-64/libInfoLogger/v2.8.3-local2/include -isystem /sw/slc9_x86-64/FairRoot/v18.4.9-alice3-local4/include -isystem /sw/slc9_x86-64/VMC/v2-1-local3/include/vmc -fPIC -O2 -std=c++20 -Werror -Wno-error=deprecated-declarations -Wno-error=unused-function -O2 -g -DNDEBUG -std=gnu++20 -fPIC -fopenmp -MD -MT Modules/TPC/CMakeFiles/O2QcTPC.dir/src/IDCs.cxx.o -MF Modules/TPC/CMakeFiles/O2QcTPC.dir/src/IDCs.cxx.o.d -o Modules/TPC/CMakeFiles/O2QcTPC.dir/src/IDCs.cxx.o -c /sw/SOURCES/QualityControl/v1.185.1/v1.185.1/Modules/TPC/src/IDCs.cxx
/sw/SOURCES/QualityControl/v1.185.1/v1.185.1/Modules/TPC/src/IDCs.cxx:26:10: fatal error: TPCBaseRecSim/CalDet.h: No such file or directory
   26 | #include "TPCBaseRecSim/CalDet.h"
```

ktf commented:
@davidrohr in the end I did not move CalDet.h... https://github.com/AliceO2Group/QualityControl/pull/2633 should fix it.
